src/actions/indexation/indexable-post-type-archive-indexation-action.php

```diff
@@ -0,0 +1,161 @@
+	 *
+	 * @var Indexable_Repository
+	 */
+	private $repository;
+
+	/**
+	 * The indexable builder.
+	 *
+	 * @var Indexable_Builder
+	 */
+	private $builder;
+
+	/**
+	 * Indexation_Post_Type_Archive_Action constructor.
+	 *
+	 * @param Indexable_Repository $repository The indexable repository.
+	 * @param Indexable_Builder    $builder    The indexable builder.
+	 * @param Post_Type_Helper     $post_type  The post type helper.
+	 */
+	public function __construct(
+		Indexable_Repository $repository,
+		Indexable_Builder $builder,
+		Post_Type_Helper $post_type
+	) {
+		$this->repository = $repository;
+		$this->builder    = $builder;
+		$this->post_type  = $post_type;
+	}
+
+	/**
+	 * The total number of unindexed post type archives.
+	 *
+	 * @return int The total number of unindexed post type archives.
+	 */
+	public function get_total_unindexed() {
+		return count( $this->get_unindexed_post_type_archives( false ) );
+	}
+
+	/**
+	 * Creates indexables for post type archives.
+	 *
+	 * @return Indexable[] The created indexables.
+	 */
+	public function index() {
+		/**
+		 * Filter 'wpseo_post_indexing_limit' - Allow filtering the amount of posts indexed during each indexing pass.
+		 *
+		 * @api int The maximum number of posts indexed.
+		 */
+		$limit = \apply_filters( 'wpseo_post_type_archive_indexation_limit', 25 );
+
+		if ( ! \is_int( $limit ) || $limit < 1 ) {
+			$limit = 25;
+		}
+
+		$unindexed_post_type_archives = $this->get_unindexed_post_type_archives( $limit );
+
+		$indexables = [];
+		foreach ( $unindexed_post_type_archives as $post_type_archive ) {
+			$indexables[] = $this->builder->build_for_post_type_archive( $post_type_archive );
+		}
+
+		return $indexables;
+	}
+
+	/**
+	 * Retrieves the list of post types for which no indexable for its archive page has been made yet.
+	 *
+	 * @param int|false $limit Limit the number of retrieved indexables to this number.
+	 *
+	 * @return array The list of post types for which no indexable for its archive page has been made yet.
+	 */
+	private function get_unindexed_post_type_archives( $limit = false ) {
+		$post_types_with_archive_pages = $this->get_post_types_with_archive_pages();
+		$indexed_post_types            = $this->get_indexed_post_type_archives();
+
+		$unindexed_post_types = \array_diff( $post_types_with_archive_pages, $indexed_post_types );
+
+		if ( $limit ) {
+			return \array_slice( $unindexed_post_types, 0, $limit );
+		}
+
+		return $unindexed_post_types;
+	}
+
+	/**
+	 * Returns the names of all the post types that have archive pages.
+	 *
+	 * @return array The list of names of all post types that have archive pages.
+	 */
+	private function get_post_types_with_archive_pages() {
+		// We only want to index archive pages of public post types that have them.
+		$public_post_types       = $this->post_type->get_public_post_types( 'object' );
+		$post_types_with_archive = \array_filter( $public_post_types, [ $this, 'is_post_type_with_archive' ] );
```

Comment:
You should be able to pass this directly to `[ $this->post_type, 'has_archive' ]`.